### ISSUE TITLE
galera-autofix failed to start after a reboot on all cluster members

### ISSUE THREAD

**Issue by nqb** [inverse-inc/packetfence#7251]
**Describe the bug**
On a customer deployment running PacketFence v12 (Debian) in a cluster, all servers have been restarted (soft reboot) at the same time. After the reboot `galera-autofix` failed to reach `packetfence-config` and then entered in a panic state.

It was necessary to restart `galera-autofix` manually for the cluster to recover.

**Expected behavior**
`galera-autofix` start correctly and fix cluster state.

**Additional context**
Logs I saw on each cluster member using `journalctl -u packetfence-galera-autofix`:

```
-- Boot 103b1c577a2c4045ac26ef3fe6b08c2c --
oct. 05 03:01:20 test-nac systemd[1]: Started PacketFence MariaDB Galera conflict resolution.
oct. 05 03:01:20 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:20+0200 lvl=info msg="Waiting 5m0s before we start checking for DB issues" pid=550
oct. 05 03:01:20 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:20+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:21 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:21+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:22 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:22+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:23 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:23+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:24 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:24+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:25 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:25+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:26 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:26+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:27 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:27+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:28 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:28+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:29 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:29+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:30 test-nac /usr/local/pf/sbin/galera-autofix[550]: t=2022-10-05T03:01:30+0200 lvl=eror msg="Cannot connect to pfconfig socket...dial tcp 127.0.0.1:44444: connect: connection refused" pid=550 function=GetLiveSeqno
oct. 05 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
oct. 05 03:01:32 test-nac galera-autofix[550]: panic: runtime error: invalid memory address or nil pointer dereference
oct. 05 03:01:32 test-nac galera-autofix[550]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6316ce]
oct. 05 03:01:32 test-nac galera-autofix[550]: goroutine 6 [running]:
oct. 05 03:01:32 test-nac galera-autofix[550]: github.com/inverse-inc/packetfence/go/pfconfigdriver.GetClusterSummary({0x70fee0, 0xc0001020c0})
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/pfconfigdriver/fetch.go:430 +0x1ae
oct. 05 03:01:32 test-nac galera-autofix[550]: github.com/inverse-inc/packetfence/go/pfconfigdriver.createQuery({0x70fee0, 0xc0001020c0}, {0x70e7e8, 0x8c2900})
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/pfconfigdriver/fetch.go:287 +0x10e
oct. 05 03:01:32 test-nac galera-autofix[550]: github.com/inverse-inc/packetfence/go/pfconfigdriver.FetchDecodeSocketCache({0x70fee0, 0xc0001020c0}, {0x70e7e8, 0x8c2900})
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/pfconfigdriver/fetch.go:346 +0x4b
oct. 05 03:01:32 test-nac galera-autofix[550]: github.com/inverse-inc/packetfence/go/galeraautofix/mariadb.DatabaseConfig(...)
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/galeraautofix/mariadb/mariadb.go:28
oct. 05 03:01:32 test-nac galera-autofix[550]: github.com/inverse-inc/packetfence/go/galeraautofix/mariadb.GetLiveSeqno({0x70fee0, 0xc000077350}, {0x6acd96, 0x9})
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/galeraautofix/mariadb/mariadb.go:132 +0xe5
oct. 05 03:01:32 test-nac galera-autofix[550]: github.com/inverse-inc/packetfence/go/galeraautofix/mariadb.GetLocalLiveSeqno(...)
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/galeraautofix/mariadb/mariadb.go:127
oct. 05 03:01:32 test-nac galera-autofix[550]: main.seqnoReporting({0x70fee0, 0xc000077350})
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/cmd/galera-autofix/main.go:87 +0x7c
oct. 05 03:01:32 test-nac galera-autofix[550]: created by main.main
oct. 05 03:01:32 test-nac galera-autofix[550]:         /builds/inverse-inc/packetfence/go/cmd/galera-autofix/main.go:32 +0x85
oct. 05 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
oct. 05 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Scheduled restart job, restart counter is at 1.
oct. 05 03:01:32 test-nac systemd[1]: Stopped PacketFence MariaDB Galera conflict resolution.
oct. 05 03:01:32 test-nac systemd[1]: Started PacketFence MariaDB Galera conflict resolution.
oct. 05 03:01:32 test-nac /usr/local/pf/sbin/galera-autofix[1213]: t=2022-10-05T03:01:32+0200 lvl=info msg="Waiting 5m0s before we start checking for DB issues" pid=1213
oct. 05 03:01:32 test-nac galera-autofix[1213]: panic: runtime error: invalid memory address or nil pointer dereference
oct. 05 03:01:32 test-nac galera-autofix[1213]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6316ce]
oct. 05 03:01:32 test-nac galera-autofix[1213]: goroutine 6 [running]:
oct. 05 03:01:32 test-nac galera-autofix[1213]: github.com/inverse-inc/packetfence/go/pfconfigdriver.GetClusterSummary({0x70fee0, 0xc00006f4a0})
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/pfconfigdriver/fetch.go:430 +0x1ae
oct. 05 03:01:32 test-nac galera-autofix[1213]: github.com/inverse-inc/packetfence/go/pfconfigdriver.createQuery({0x70fee0, 0xc00006f4a0}, {0x70e7e8, 0x8c2900})
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/pfconfigdriver/fetch.go:287 +0x10e
oct. 05 03:01:32 test-nac galera-autofix[1213]: github.com/inverse-inc/packetfence/go/pfconfigdriver.FetchDecodeSocketCache({0x70fee0, 0xc00006f4a0}, {0x70e7e8, 0x8c2900})
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/pfconfigdriver/fetch.go:346 +0x4b
oct. 05 03:01:32 test-nac galera-autofix[1213]: github.com/inverse-inc/packetfence/go/galeraautofix/mariadb.DatabaseConfig(...)
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/galeraautofix/mariadb/mariadb.go:28
oct. 05 03:01:32 test-nac galera-autofix[1213]: github.com/inverse-inc/packetfence/go/galeraautofix/mariadb.GetLiveSeqno({0x70fee0, 0xc00006f350}, {0x6acd96, 0x9})
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/galeraautofix/mariadb/mariadb.go:132 +0xe5
oct. 05 03:01:32 test-nac galera-autofix[1213]: github.com/inverse-inc/packetfence/go/galeraautofix/mariadb.GetLocalLiveSeqno(...)
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/galeraautofix/mariadb/mariadb.go:127
oct. 05 03:01:32 test-nac galera-autofix[1213]: main.seqnoReporting({0x70fee0, 0xc00006f350})
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/cmd/galera-autofix/main.go:87 +0x7c
oct. 05 03:01:32 test-nac galera-autofix[1213]: created by main.main
oct. 05 03:01:32 test-nac galera-autofix[1213]:         /builds/inverse-inc/packetfence/go/cmd/galera-autofix/main.go:32 +0x85
oct. 05 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

**Comment by julsemaan** [comment]
Technically, it should have kept restarting on failure but there must be a systemd parameter we need to put so that there is no limits to the restarts

**Comment by julsemaan** [comment]
Looks like this answers it: https://serverfault.com/a/753444

We need to tweak the systemd settings for that.

Also using monit as a safety net should have worked. Was monit enabled on this environment?

**Comment by nqb** [comment]
> Also using monit as a safety net should have worked. Was monit enabled on this environment?

Monit was **not** enabled on this setup.

You are certainly right regarding solution. Logs I see in syslog:
```
Oct  5 03:00:39 test-nac systemd[1]: packetfence-galera-autofix.service: Succeeded.
Oct  5 03:00:39 test-nac systemd[1]: packetfence-galera-autofix.service: Consumed 4min 14.757s CPU time.
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Scheduled restart job, restart counter is at 1.
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Scheduled restart job, restart counter is at 2.
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  5 03:01:32 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
Oct  5 03:01:33 test-nac systemd[1]: packetfence-galera-autofix.service: Scheduled restart job, restart counter is at 3.
Oct  5 03:01:33 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  5 03:01:33 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
Oct  5 03:01:33 test-nac systemd[1]: packetfence-galera-autofix.service: Scheduled restart job, restart counter is at 4.
Oct  5 03:01:33 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  5 03:01:33 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
Oct  5 03:01:34 test-nac systemd[1]: packetfence-galera-autofix.service: Scheduled restart job, restart counter is at 5.
Oct  5 03:01:34 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  5 03:01:34 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
Oct  5 03:01:34 test-nac systemd[1]: packetfence-galera-autofix.service: Scheduled restart job, restart counter is at 6.
Oct  5 03:01:34 test-nac systemd[1]: packetfence-galera-autofix.service: Start request repeated too quickly.
Oct  5 03:01:34 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
[..]
Oct  5 09:04:05 test-nac systemd[1]: packetfence-galera-autofix.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  5 09:04:05 test-nac systemd[1]: packetfence-galera-autofix.service: Failed with result 'exit-code'.
```

=> systemd didn't try more than 5 times.

**Comment by julsemaan** [comment]
Looked at the doc and adding `StartLimitInterval=0` disables the rate-limiting

Tested it on my machine and galera-autofix kept restarting over and over

Fix backported until 10.3
